tests/spec/core/dfn-spec.js

```diff
@@ -329,4 +329,218 @@ describe("Core — Definitions", () => {
+      </section>
+      `;
+      const ops = makeStandardOps(null, body);
+      const doc = await makeRSDoc(ops);
+      const dfn = doc.querySelector("#abstract-op dfn");
+      expect(dfn.dataset.dfnType).toBe("abstract-op");
+      expect(dfn.dataset.export).toBe("");
+    });
+
+    it("handles attributes", async () => {
+      const body = `
+        <section>
+          <h2>Attributes</h2>
+          <p id="attributes">
+            <dfn class="attribute">some-attribute</dfn>
+          </p>
+          <p id="attribute-bad">
+            <dfn class="attribute">-attribute</dfn>
+          </p>
+        </section>
+      `;
+      const ops = makeStandardOps(null, body);
+      const doc = await makeRSDoc(ops);
+      const dfn = doc.querySelector("#attributes dfn");
+      expect(dfn.dataset.dfnType).toBe("attribute");
+      expect(dfn.dataset.export).toBe("");
+
+      // Check validation error
+      const errors = doc.respec.errors.filter(err => err.plugin === "core/dfn");
+      expect(errors).toHaveSize(1);
+    });
+
+    it("handles attribute values", async () => {
+      const body = `
+        <section>
+          <h2>Attribute values</h2>
+          <p id="attr-value">
+            <dfn class="attr-value" data-dfn-for="input/type">password</dfn>
+          </p>
+          <p id="attr-value-errors">
+            <dfn class="attr-value">some-text</dfn>
+            <dfn class="attr-value">-not-ok!</dfn>
+          </p>
+        </section>
+      `;
+      const ops = makeStandardOps(null, body);
+      const doc = await makeRSDoc(ops);
+      const dfn = doc.querySelector("#attr-value dfn");
+      expect(dfn.dataset.dfnFor).toBe("input/type");
+      expect(dfn.dataset.dfnType).toBe("attr-value");
+      expect(dfn.dataset.export).toBe("");
+
+      // Check validation error
+      const errors = doc.respec.errors.filter(err => err.plugin === "core/dfn");
+      expect(errors).toHaveSize(2);
+    });
+
+    it("handles elements", async () => {
+      const body = `
+        <section>
+          <h2>Elements</h2>
+          <p id="elements">
+            <dfn class="element">element</dfn>
+          </p>
+          <p id="elements-bad">
+            <dfn class="element">-element</dfn>
+          </p>
+        </section>
+      `;
+      const ops = makeStandardOps(null, body);
+      const doc = await makeRSDoc(ops);
+      const dfn = doc.querySelector("#elements dfn");
+      expect(dfn.dataset.dfnType).toBe("element");
+      expect(dfn.dataset.export).toBe("");
+
+      // Check validation error
+      const errors = doc.respec.errors.filter(err => err.plugin === "core/dfn");
+      expect(errors).toHaveSize(1);
+      expect(errors[0].message).toContain("-element");
+    });
+
+    it("handles element states", async () => {
+      const body = `
+        <section>
+          <h2>Element states</h2>
+          <p id="element-state">
+            <dfn data-dfn-for="input" data-dfn-type="element-state">Password</dfn>
```

Comment:
I think we're supposed to be using `class="element-state"` here?

Reply:
oops, yes.